### ISSUE TITLE
[TimePicker] More natural 24h hour picker

### ISSUE THREAD

**Issue by hrach** [material-components/material-components-android#1450]
**Is your feature request related to a problem? Please describe.**

I don't have much "experience" with the new 24h hour picker in "one" circle. Though I feel this to be pretty unnatural (an European user here). 

**Describe the solution you'd like**

Bring back the system's two inner & outer circles.

**Describe alternatives you've considered**

I and users will get used to.

**Additional context**

![time_picker](https://user-images.githubusercontent.com/284263/86094782-47838200-bab1-11ea-81c3-d57b989725a4.png)

**Comment by matpag** [comment]
Actually in the official Clock application in my Pixel 3a, there are 2 circles, the outer one for 0-12 hours and the inner one for 13-00. (I'm in EU too).

I agree we should have the standard UI for timepicker in 24H format.

**Comment by angdesign** [comment]
Hi hrach, we moved to a single ring design to prioritize a11y considerations for a minimum touch target. In addition, to account for responsive ranges, the dual ring proved to be problematic at smaller viewport sizes when the ring begins to constrict. So this simplified design is more resilient, but noted that it may be unfamiliar.

**Comment by matpag** [comment]
This is a good (and valid) explanation, thank you

**Comment by hrach** [comment]
@angdesign thanks for the info.

> but noted that it may be unfamiliar.

Actually, this is something which contradicts with how is the time perceived. It not a new concept (~unfamiliar), it is redesigned already existing one (~unnatural). 

I understand the reasoning behind, though not sure if current solution is good. Hopefully you have some research for any variant. I have opened this just after my first impression. :) I'm still not much convinced that smaller displays (in 2020) / a11y are good reason ending with this specific solution.

**Comment by angdesign** [comment]
Feedback noted hrach, thanks. And agreed that there is the unnatural trade-off with this new direction. We have data to support this direction from an internal study on time pickers, however, with that said we will also monitor feedback on this for future considerations.

**Comment by brudaswen** [comment]
Actually the fact that only every second number is visible confused me even more and makes it hard/unnatural to select 13 for example.

**Comment by mattlaabs** [comment]
I was excited about the new time picker, but the 24h layout is an absolute show stopper. I would never publish this to a European user base. Please at least make this optional, so app developers can choose to maybe use this on very small devices, but even then I'd rather fallback to keyboard mode.

**Comment by tipa** [comment]
Switched to the new MaterialTimePicker last week and just got the 2nd negative (1-star) review because of the new 24h layout.
I understand and agree with the reasoning for the change, but my users don't care 🤷‍♂️

**Comment by ohmsweetohm1** [comment]
I think the main reason I don't like the new design is that it does not mirror a real wall clock anymore. We all have years of experience of reading the time of a clock and expect the e.g. 6 to be on the bottom of any clock.

**Comment by lukasz-kalnik-gcx** [comment]
@angdesign I see your point regarding touch target size for the inner circle.
However the current solution is very misleading and I would argue it has even worse a11y, especially for non-tech savvy users.

1. If the time picker is meant to mimic a clock face, there are no circular 24 hour clocks. No one has ever seen a clock which on its face has every second hour (0, 2, 4, 6...).
2. Mimicking the clock face works well for 12 hour format, but not for 24 hours. So the 24 hour picker seems like an afterthought and not something that was considered from the beginning of the design.
3. 24 hour format is normally exclusively used in digital clocks which just display the time as digits (don't have a round clock face with numbers).
4. If a user wants to select say 13:00 he/she doesn't know where to tap. This was misleading even for me as a technical user, but even more so for older/non-technical users.

So to sum up, for a proper time picker for 24 hour format with good a11y for everyone probably just a scrolling list of numbers (like iOS is doing) would be much better solution.

I know this is a complete break from the existing Material Design guidelines, but I doubt any user will like this half baked solution.

What's worse, the [official Material guidelines](https://material.io/components/time-pickers) don't even show any example of a 24 hour Time Picker. It really looks like it was just forgotten from the beginning. Not all the world uses 12 hour time notation...

**Comment by lukasz-kalnik-gcx** [comment]
BTW something more regarding the touch target size.
If you try to select say 13:00 you have to tap between the numbers 12 and 14 on the clock face, it's also a very small touch target.
If you want to first tap say 12 and drag the pointer to 13, you cannot do it, because after you selected an hour the clock immediately changes to the minutes selection mode. So you have to go back to the hour selection mode and reposition the hour marker.

**Comment by petrnalevka** [comment]
Hello, I would like to contribute to this debate. I have implemented the new Material Design Time picker in my alarm clock app Sleep as Android (+1M active users, 20M downloads) and got storm of incredibly bad feedback on the 24-hour picker variant. 

My users feel it is absolutely unnatural to select 12:00 noon at the bottom of the circle - specially for an alarm clock where you usually select the "AM" times the clock becomes completely distorted. Not only the position of hours which does not respect traditional position on the clock but also some users had issues to understand that even times are hiding between the odd numbers. 

I just had it in a 5% staged roll out but the bad feedback was so massive that I had to immediately reverse to the old time picker. 

Here is one of the threads I got from my users https://forum.urbandroid.org/t/give-us-back-analog-clock-picker/7513/4

Than I looked into the Google official apps to realize that all of them still use the old picker. So I guess this is for a reason?

Is there any plan to fix this? As I think nobody can seriously use the 24-hour variant of the new material design time picker in an app that makes the use of a time picker an every day task for the users..

Many thanks..

**Comment by jmartingit** [comment]
In my opinion, if we're not going to follow the conventional positions on the clock, a linear slider would probably be more intuitive.

I've always thought the time picker from the old (and Google-owned) 'Timely' alarm app was faster and more intuitive than any round picker (for both 24-hour and am/pm modes).

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
I am believing that regardless of what the picker shall become, the default time should merely be typed.

![Screenshot_20210911-191325](https://user-images.githubusercontent.com/42837531/132959015-d6a74f74-6a30-44fd-8c26-715b72b53238.png)

Surely this design is more adaptive, and is less likely to be modified than a picker that may have rings that may become removed, [that may become a slider](http://github.com/material-components/material-components-android/issues/1450#issuecomment-852860928) , and/or that may have rings retroactively added after removal. It is also much more easy to understand.

**Comment by Skwara** [comment]
Hello,

I would like to add my opinion to this subject.

Generally, I think this design is fundamentally flawed. It breaks one of the core reflex related to working with the clock. When I want to sent a reminder to 13:00 (1 PM), I immediately look at the top of the clock face. With this design is quite the opposite.

This is not something we can simply get used to. Clocks exist outside of Android and they don't have 1 PM at the bottom. This will result in a constant confusion. Let's add to that the system fragmentation, where not every app will embrace this new design. We will be constantly confused not only between real and virtual worlds, but also inside the virtual one. There is no chance for working out a reflex unless someone will only use apps that utilize one clock style.

I can understand that someone from US might completely miss this. There is no way he or she could understand the problem as it's unique to people using the 24h time format. If I have to compare it to something, imagine that in the new material design, US users are forced to use the metric system. It would be unnatural to them, fragmentation will be huge, just like in this case. Or if you suddenly change the text direction to right to left. Honestly, these lie in the same category as the clock idea to me.

The solution with a manual input is simply worse. The previous clock style was perfect. Two taps and I'm done. Now it's a tap on the little keyboard icon, another tap on the hour field, another tap on the hour field because the keyboard doesn't pop up the first time (but maybe it's unique to me), one or two taps for the hour, another tap to move to minutes and again two taps for minutes. This is so much worse than the previous solution.

Technology should evolve for humans, not against them, just because there is not enough room for virtual buttons. The suggestion that users will get used to it defeats the purpose of evolving the technology, and in this case I don't even see how users will get used to this as I explained at the beginning of this post.

Thank you and please revert this.

**Comment by Skwara** [comment]
> > I can understand that someone from US might completely miss this. There is no way he or she could understand the problem as it's unique to people using the 24h time format.
> 
> You are naïve. 😄 Do you actually think, that people working for google are not smart enough to understand cultural differences? They just don't care and try to enforce their culture on us. It's just the mentality of the culture they are living in. Their culture doesn't care or respect foreign cultures, even the one which has born their own culture.

I wrote that I **understand** that someone from US might **miss** this. This means that I can imagine a scenario, where this happens, but I'm not claiming it's the most probable one.

Your assumption that people working on UI in Google are smart enough to understand all cultural differences is not based on facts, it's rather a belief based on the "it's Google, right?" idea. Don't forget, that Google is just another big corporation. Things fall between the cracks even in Google or Apple.

Saying that they simply doesn't care is just being negative. You don't know that and it doesn't bring anything to the discussion. It definitely doesn't help.

Let's focus on the subject. Let's try to describe how you feel about the issue. What's wrong with it. What would be a good solution. Let's help the decisive people understand the nature of the problem. They know but they can't feel what we feel if they haven't been using 24h time format throughout their whole lives. Let's help them understand.

**Comment by Skwara** [comment]
> Hi folks,
> 
> Thanks for providing your opinions and sharing your feelings. We also got a lot of feedbacks internally. As far as I know our design team is aware of the issue and have a research/redesign plan to figure out a better solution of 24h format.
> 
> I'm sorry that as an eng I cannot promise a timeline here (we need to wait for a design decision before we can really do anything) but I totally understand your feelings and will try my best to keep tracking on this to prevent it falling through the cracks.
> 
> Feel free to provide more of your thoughts on this or any ideas that we can present to our design team.

If I have to be honest, I can't imagine a better idea than the previous, double circle one. It's simple and does the job. I think it's key advantages are:

1. Keeping with the natural alignment of the clock.
2. Requiring only two taps to choose a time up to 5 minute accuracy. More accuracy is also possible with dragging.

I can't emphasize enough how much these two things are important to me and I believe should be the baseline for any new idea.
If the new idea breaks core reflexes, we already talked about it. If the new idea will require more taps to achieve the same thing, what's the point?

That's why my idea is to stick to the old approach. If that won't be possible because it has to be new just for the sake of being new, at least it shouldn't break those two points I mentioned above.

Another idea would be to have a normal clock's face with a button in the center that would switch between AM and PM hours around the clock. It's simply worse than the previous one because it breaks point 2 but at least it implements point 1. It has a potential for minimizing the frequency of taps on that center switch by predicting if a user needs AM or PM times but on the other hand, it removes the ability to train a reflex, where a user would always know that he/she has to tap the switch first to see the PM times.

The best of both worlds would be to have an ability to choose which style a user wants. Make the material one the default but have the option to switch to the other one. I never understood why a choice became so passé these days and users needs to adapt instead of deciding.

**Comment by drchen** [comment]
Hi @Skwara,

Thanks for your detailed explanation. I think the two key advantages you mentioned is quite convincing, at least to me personally. I'll raise them to our designers when we have a discussion. 

Again, thanks everyone for your input. They are all very valuable feedbacks to our upcoming design plans. : )

**Comment by Skwara** [comment]
I would also like to underline the lack of consistency that exists since the introduction of the new clock style.

I use Keep, Calendar and Gmail to handle my notes, appointments and emails. I have to use the new style in Keep but the old one in Calendar and Gmail. This is quite irritating as I'm unable to any one of those. What was the point of introducing it only in Keep?

Can you revert this change in Keep please?

**Comment by Skwara** [comment]
@drchen Do I understand correctly, that you decided to bring back the double circle approach?

**Comment by drchen** [comment]
Yes, correct. : )

**Comment by Skwara** [comment]
@drchen Today I was greeted with a new time picker approach. True, the double circle approach is back but I am very disappointed that it was pushed to be the non-default approach. The default is direct input from keyboard. I think this is not an optimal approach for 2 main reasons, that I describe below:

1. The keyboard approach requires more input from the user to select the time. Let me compare (I assume the time picker dialog is already open):
   - Keyboard:
      - tap into the hour field even though the cursor is already there to force the keyboard to pop up,
      - 1-4 taps to provide the time (1 tap for 5 AM, because minutes are default to 00, 4 taps for 13:30 to provide all 4 digits).
   - Double circle:
      - tap for hour,
      - tap for minute (up to 5 min accuracy) or tap + drag (for 1 minute accuracy).

   In comparison, the double-circle approach requires less taps to provide the time up to 5 minute accuracy and almost always less effort to provide the time for 1 minute accuracy. The only scenario where keyboard is better is when the user really needs the one minute accuracy for times before 10:00 AM (2 taps for keyboard and 1 tap + 1 tap and drag for double circle) which is barely any win here. This brings me to my second reason:

2. The default approach doesn't fit the use case. The keyboard approach is (arguably, because of the drag option) better for providing times up to 1 minute accuracy. You simply type it compared to dragging with double circle one. The problem here is I believe in most use cases this is not necessary. For reminders and e-mail snoozing 5 minute accuracy is enough as I believe most times provided for those are full hour (XX:00) or half hour (XX:30). I think reminders like XX:27 are rarely needed.
Regardless of that, it was decided to put keyboard as the default method that compromises ease-of-use for accuracy that is not required most of the time. **It doesn't use the best thing about the double circle approach which is requiring at most 2 taps for any hour up to 5 minutes accuracy.** I don't understand this decision.

Given the above, can you answer the following?
1. Is it possible to allow the user to choose the default method globally, for the entire system?
2. Why it was decided to put keyboard input as default? Am I missing something?

**Comment by petrnalevka** [comment]
Many thanks for reverting to the 2-rings picker for 24-hour format. I see the change made in flutter. Any estimate on when this will be available in the material-components-android library? I see this is not yet part of the 1.7.0-beta01. Many thanks for looking into this issue..

**Comment by paulfthomas** [comment]
@Skwara thanks for the feedback:

1. No, there is no system-wide setting for this unfortunately. It is still possible though to change the default input mode using the Builder: `MaterialTimePicker.Builder().setInputMode(INPUT_MODE_CLOCK)`. An app can easily offer an option to save the user preference.

2. Our research indicated a preference for text input mode when using the 24h format. It is less confusing and more accurate for the majority of users.

**Comment by paulfthomas** [comment]
@petrnalevka this will be available in 1.8.0

**Comment by Skwara** [comment]
@paulfthomas Thank you for your responses. Would it be possible to at least make the keyboard pop-up automatically when the time picker is being displayed? I don't see a reason to require the user to activate it manually.

I think I can also assume I should drop any hopes that Google Keep, Gmail etc. will utilize the `Builder` method to allow the user to choose.

**Comment by woj-tek** [comment]
> Our research indicated a preference for text input mode when using the 24h format. It is less confusing and more accurate for the majority of users.

Could you share more light on this? How and where did you perform the research? Did those using 24h were "24h natives" or simply folks from countries normally using am/pm and force to user 24h system? I know it's anectotal but I know noone that would prefer text-input in that case using normally 24h time...

**Comment by BEEDELLROKEJULIANLOCKHART** [comment]
I really like that the new version uses the keyboard by default, but indeed I want the keyboard to appear and the field be focussed automatically.

**Comment by hrach** [comment]
> Our research indicated a preference for text input mode when using the 24h format. It is less confusing and more accurate for the majority of users.

I already heard this argument two years ago when I created the issue, and it turned out to be major nonsense anyway.

**Comment by Skwara** [comment]
I want to underline that the current solution is still inconsistent with other Google applications. In Gmail I still get the double circle by default and a love it. It will still hurt my brain to have to constantly switch time input methods. Please make a global switch where anyone can choose the default method that would be global for at least google apps.

**Comment by Skwara** [comment]
There's no activity here any more and I am afraid it might stay like that, so just w few final words from myself:

@paulfthomas I understand that it might be frustrating when you made effort and changed something about the thing people were "whining" about and they are still not satisfied. You might think "people will never be satisfied, why even care?".

On the other hand I have to underline, that to me, personally, the change doesn't make sense and it's for worse. I have reasons:
1. If you concluded that keyboard input is better, why the keyboard is not presented to the user? Looks like a major oversight.
2. If keyboard input is better, why Gmail and Calendar use the double wheel? Don't get me wrong, I love it and I hope it will stay that way as long as possible but I hate the inconsistency. Even a perfect solution is wrong when the user is forced to use it next to other approaches to the same problem. The inconsistency hurts my brain because I am unable to adapt. It was underlined many times in this thread.
3. If you decided to allow the user to choose, you have to see the absurdity of forcing the user to choose every time the feature is used. It's a matter of global preference, not each time the feature is used. If someone prefers double circle, he will prefer it every time so forcing him to click the icon every time is pointless and irritating.

And finally what's the point of this discussion, inputs from users etc. if at the end you throw the conclusions away and do it your way? There were reasons, there were votes and the majority of people liked the double circle approach. Does it matter?

@rokejulianlockhart yes, I see that you strongly prefer the digital input. It makes sense but not fully in the current form. Also you have to observe that you are in minority.

Sorry for being negative here but I'm struggling to see the logic in what happened. I would like to find it.

**Comment by drchen** [comment]
Hi thanks for your feedback and I want to highlight one thing - the decision has been made via an extensive UX research, instead of just observing a few thumbs-up/down on a github comment. : )

I have to admit I personally was also surprised by the UXR result - but that's why we need UXR because sometimes the large audience's preference can be just opposite to our personal intuition or observation.

As for global settings - unfortunately we are just a UX library and we have no control over any apps. Providing the option for each app to opt-in/out the behavior is already the best we can do. Thanks for your understanding.

**Comment by paulfthomas** [comment]
@Skwara no need to apologize, user feedback is great for us, the goal is to build components that people like to use! I wish we could fix/improve the code faster, I know it can be frustrating for the devs using our library, we always try to prioritize some of the reported issues along with building new features.

I can maybe try to answer few of your questions:

1. The keyboard should be presented to the user, this is a current bug we're looking into, if you look at the logic you will see that the method is called but nothing happens when the picker is first shown. This is definitely not an oversight and will be fixed soon I hope. A PR is also very welcome ;)
2. The other apps have their own release cycle, they will likely migrate to the Material version, a possibility would also be to add an app setting letting the user decide on the input type or save the last used one. I don't have more information but this is something that will be looked into.
3. I can see a system setting could be nice, I also believe that an app level setting/remembering the last choice is probably a good option too, the picker can be used in different contexts and a user may prefer text input vs double circle in some instances.

Hope this helps!

**Comment by Skwara** [comment]
@drchen @paulfthomas Thank you very much for your input.

You're right @drchen than the UXR is far more reliable than this thread, no doubt about it. Thanks for underlying that.

Thank you for your explanations @paulfthomas, it's now clear to me what's happening and it makes a lot of sense now.

Thank you for your efforts.

**Comment by levsha** [comment]
The previous decision to use single 24 hours circle was also based on UXR. What happened then? Majority of users changed their mind dramatically?
Sorry, your claim "the decision has been made via an extensive UX research" means just nothing, especially without proving any objective confirmations.

**Comment by drchen** [comment]
I'm not sure if the previous decision was based on UXR result or not - for what I understand it's mostly because of accessibility requirements.

If you have any better or more "objective" suggestions than trusting UXR result to make those decisions, they will be more than welcome. : )

**Comment by woj-tek** [comment]
> If you have any better or more "objective" suggestions than trusting UXR result to make those decisions, they will be more than welcome. : )

@drchen I think it was suggested to maybe shere the data and how the research was conducted :-) I already asked @paulfthomas about that previously (https://github.com/material-components/material-components-android/issues/1450#issuecomment-1231796684) but got no response. What's more - if the research is badly specified then the outcome would yield incorrect results :-)

**Comment by drchen** [comment]
I cannot represent the UXR team to speak here but I feel there's a high chance that those info are confidential. 

Maybe @paulfthomas can add more details here.

**Comment by levsha** [comment]
UXR is not something that Google (or any other IT company) introduced. Many companies in real world do this, and actually have to do this. Imagine if a pharmaceutical company would just claim "We did research and can confirm this new medicine works! No we won't share details with you". There are well known protocols existing since decades how to publish research raw data without disclosure of private data.
Sorry, but again - without objective data your claims "it's all based on UXR" are just claims "trust me"

**Comment by levsha** [comment]
"The other apps have their own release cycle, they will likely migrate to the Material version" - native "Alarm" app on Pixel phones, time picker on Google Calendar - they all have been always using two circles time picker. I just checked - now they have the little keyboard icon on the bottom left that also allows to set the time using keyboard if you wish. Maybe ask Pixel or Calendar teams (better an European branch) for their UXR results? They should definitely have them.

**Comment by paulfthomas** [comment]
As @drchen mentioned, the single 24h circle was originally picked for accessibility reasons, it leaves more room for bigger touch targets/no overlapping. It was unfortunately a mistake that we are trying to fix with the redesigned double circle version. Several apps are still using the 24h framework version and have been waiting on this redesign to adopt the Material version. I cannot share the UXR results but "24h natives" were definitely involved, I am one myself :)

**Comment by Skwara** [comment]
The keyboard started sliding out automatically. Thank you guys for the fix :*

**Comment by levsha** [comment]
I have Google Pixel 7 now with all latest updates. And behavior is still inconsistent: Google Calendar, Alarm Clock use the widget that show analog clock immediately, Google Keep shows the widget with digital input first, without any way to change this default.

**Comment by levsha** [comment]
Two years after this issue being closed, the mistake was still not acknowledged and not reverted. 
And today on any supported Google Pixel phone with the latest Android and latest updates you have two circles time picker on native Clock app when you edit alarms, you have the same kind of two circles time picker in Google Calendar app, but you still have keyboard input type time picker on apps using this code, like Google Keep.
Is there UX Research maybe that confirms that users like inconsistent UI?
